tfhe/src/integer/gpu/list_compression/server_keys.rs

```diff
@@ -0,0 +1,328 @@
+                cuda_memcpy_async_gpu_to_gpu(
+                    dest_ptr,
+                    ciphertext.d_blocks.0.d_vec.as_c_ptr(gpu_index),
+                    size as u64,
+                    streams.ptr[gpu_index as usize],
+                    streams.gpu_indexes[gpu_index as usize],
+                );
+
+                offset += ciphertext.d_blocks.0.d_vec.len;
+            }
+
+            streams.synchronize();
+            d_vec
+        };
+
+        let flattened_ciphertexts =
+            CudaLweCiphertextList::from_cuda_vec(d_vec, lwe_ciphertext_count, ciphertext_modulus);
+
+        let info = vec_ciphertexts
+            .iter()
+            .flat_map(|x| x.info.blocks.clone())
+            .collect_vec();
+
+        (flattened_ciphertexts, info)
+    }
+
+    pub fn compress_ciphertexts_into_list(
+        &self,
+        ciphertexts: &Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+        gpu_index: u32,
+    ) -> (
+        CudaGlweCiphertextList<u64>,
+        Vec<CudaBlockInfo>,
+        CiphertextModulusLog,
+    ) {
+        let lwe_pksk = &self.packing_key_switching_key;
+
+        let ciphertext_modulus = lwe_pksk.ciphertext_modulus();
+        let compress_polynomial_size = lwe_pksk.output_polynomial_size();
+        let compress_glwe_size = lwe_pksk.output_glwe_size();
+
+        let first_ct = ciphertexts.first().unwrap();
+        let first_ct_info = first_ct.info.blocks.first().unwrap();
+        let message_modulus = first_ct_info.message_modulus;
+        let carry_modulus = first_ct_info.carry_modulus;
+
+        let lwe_dimension = first_ct.d_blocks.lwe_dimension();
+
+        let num_lwes: usize = ciphertexts
+            .iter()
+            .map(|x| x.d_blocks.lwe_ciphertext_count().0)
+            .sum();
+
+        let mut output_glwe = CudaGlweCiphertextList::new(
+            compress_glwe_size.to_glwe_dimension(),
+            compress_polynomial_size,
+            GlweCiphertextCount(ciphertexts.len()),
+            ciphertext_modulus,
+            streams,
+        );
+
+        let (input_lwes, info) = Self::flatten(ciphertexts, streams, gpu_index);
+
+        unsafe {
+            compression_compress_integer_radix_async(
+                streams,
+                &mut output_glwe.0.d_vec,
+                &input_lwes.0.d_vec,
+                &self.packing_key_switching_key.d_vec,
+                message_modulus,
+                carry_modulus,
+                GlweDimension(0),
+                PolynomialSize(0),
+                compress_glwe_size.to_glwe_dimension(),
+                compress_polynomial_size,
+                lwe_dimension,
+                lwe_pksk.decomposition_base_log(),
+                lwe_pksk.decomposition_level_count(),
+                self.lwe_per_glwe.0 as u32,
+                self.storage_log_modulus.0 as u32,
+                num_lwes as u32,
+            );
+        }
```

Comment:
You need to synchronize, otherwise this function should be unsafe and have an `_async` suffix.

Reply:
Indeed.

I converted `flatten` to `flatten_async`.